staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go

```diff
@@ -101,55 +67,100 @@ func WithAudit(handler http.Handler, requestContextMapper request.RequestContext
 			responsewriters.InternalError(w, req, errors.New("no context found for request"))
 			return
 		}
-		attribs, err := GetAuthorizerAttributes(ctx)
-		if err != nil {
-			responsewriters.InternalError(w, req, err)
-			return
+
+		// prefer the id from the headers. If not available, create a new one.
+		// TODO(security): do we want to forbid the header for non-front-proxy users?
+		var id string
+		ids := req.Header[AuditIDHeader]
+		if len(ids) > 0 {
+			id = ids[0]
+		} else {
+			id = uuid.NewRandom().String()
 		}
 
-		username := "<none>"
-		groups := "<none>"
-		if attribs.GetUser() != nil {
-			username = attribs.GetUser().GetName()
-			if userGroups := attribs.GetUser().GetGroups(); len(userGroups) > 0 {
-				groups = auditStringSlice(userGroups)
+		// set the level
+		level := auditinternal.LevelMetadata
+		if policy != nil && len(policy.Rules) > 0 {
+			// This is just a hack to get through the test without setting a high level by default.
+			// TODO: add the policy evalutation here
+			level = policy.Rules[0].Level
+		}
+		headerLevels := req.Header[AuditLevelHeader]
+		if len(headerLevels) > 0 {
+			switch l := auditinternal.Level(headerLevels[0]); l {
+			case auditinternal.LevelMetadata, auditinternal.LevelNone, auditinternal.LevelResponse, auditinternal.LevelRequest:
+				level = l
+			default:
+				glog.Warningf("Invalid audit level in %v helper: %q", AuditLevelHeader, l)
 			}
 		}
-		asuser := req.Header.Get(authenticationapi.ImpersonateUserHeader)
-		if len(asuser) == 0 {
-			asuser = "<self>"
+
+		ev := &auditinternal.Event{
+			Level:     level,
+			Timestamp: metav1.NewTime(time.Now()),
+			AuditID:   types.UID(id),
+		}
+		ctx = request.WithAuditEvent(ctx, ev)
```

Comment:
The context could be used in multiple go routines, but the handlers in rest.go are linear, with a final `WriteObjectNegotiated`. It wouldn't be healthy if we have concurrent response writers. 

As we only write twice per request though, we could switch it to by-value without much overhead. Then we are on the safe side. But IMO it's not necessary.